### ISSUE TITLE
Examples of memory leaks fixed by the adoption of generation checks

### ISSUE THREAD

**Issue by fabiospampinato** [proposal-signals/proposal-signals#106]
As I understand it the `generation` hidden variable in this proposal exists precisely to be able to know if a computed could be dirty or not by comparing it against the `generation` of the signals it depends on, when needed, since if a computed is not watched the signals it depends on won't hold a reference to the computed anymore, so the signals won't be able to directly tell the computed that it might be dirty.

As I understand it the idea is that by trading some performance to keep around and update these counters one gets the ability to garbage-collect these computeds.

I'm not sure about the trade-offs here, I think it would be good to provide more details in this regard, in particular: 

1. Are we sure this is better than making those references from the signals to the computeds weak references instead?
2. Can some example code where this implementation would actually enable some computed to be garbage-collected automatically be added to the proposal for extra clarity? This seems one of the key wanted behaviors that some important decisions seem based upon, and just thinking about the code where I use signals I can't actually think of a single scenario where I'm leaking memory this way, nor I'm sure what that scenario could look like.
3. Is this really actually about the lack of an auto-disposal mechanism for computes and watchers? Because thinking about a modern signals-only framework like Solid I'm not actually sure these kinds of memory leaks exist there. It would be good to know if this is meant to address an issue that all signals frameworks are subjected to or not, to better evaluate the trade-offs.

**Comment by littledan** [comment]
Yes, I agree that it will be important to improve documentation here. The goal is to enable separation of the view from the model. If computeds always have to be owned by part of the view (for auto-disposal), the separation is less flexible. This vague claim that I just made should be substantiated with a concrete code example.

**Comment by elliotgoodrich** [comment]
I have an issue suggesting that we use weak references internally (hopefully if signals is implemented natively we do not have to pay the cost of `WeakRef`) because of certain situations where memory leaks can occur https://github.com/tc39/proposal-signals/issues/265 - is this the sort of thing you're looking for @fabiospampinato?

**Comment by fabiospampinato** [comment]
@elliotgoodrich potentially yes, if that's what the readme was referring to. 

That kind of situation you mentioned seem artificial to me though, as far as I've seen, as in: one can construct the type of situation you mention in your post, for sure, but in practice that doesn't really happen in modern frameworks, because if you make a memo you might use it if not now maybe in the future, so disposing of it prematurely just because you are not using it right now may even be considered a bug, and when for sure that memo will never be used again it's when its parent computation gets refreshed, in which case the memo will be destroyed automatically, so a situation where you have a memo that leaks memory indefinitely doesn't really exist, one would need to create a top-level memo that nobody has a reference to for some reason, it just doesn't happen.

On the other hand the cost of using weak refs is definitely real, so the tradeoff to me here seems giving up something real in order to gain something that is not actually useful in practice, probably.

**Comment by elliotgoodrich** [comment]
It doesn't seem as though hierarchical UI libraries have too much of an issue being able to manually unwatch signals.  However, the issue I opened was from the context of non-UI libraries, where the lifetime of signals is not so easily tracked and their lifetimes mirror the lifetimes of some other object held the application.

If there is no good and fast solution to my issue, then I don't think it will be a blocker to the Signals proposal, but if it could be solved without any drawback (performance or otherwise) then it would make Signals applicable in non-UI contexts.

So in direct response to your questions

> 1. Are we sure this is better than making those references from the signals to the computeds weak references instead?

No we aren't sure.  Much more investigation will be needed, most likely requiring us to implement Signals within the browsers themselves. Since my own personal experiments with using `WeakRef` or a native C++ node module had other performance implications.

> 2. Can some example code where this implementation would actually enable some computed to be garbage-collected automatically be added to the proposal for extra clarity? This seems one of the key wanted behaviors that some important decisions seem based upon, and just thinking about the code where I use signals I can't actually think of a single scenario where I'm leaking memory this way, nor I'm sure what that scenario could look like.

I think examples are given in https://github.com/tc39/proposal-signals/issues/265. Even though they may be artificial, I am personally experiencing the first 2 in integrating Signals with a non-UI library, and @DavidANeil (I think) was worried about the 3rd issue in his library.

> 3. Is this really actually about the lack of an auto-disposal mechanism for computes and watchers? Because thinking about a modern signals-only framework like Solid I'm not actually sure these kinds of memory leaks exist there. It would be good to know if this is meant to address an issue that all signals frameworks are subjected to or not, to better evaluate the trade-offs.

Yes these leaks are hard inside modern hierarchical frameworks, but they can easily occur in non-hierarchical libraries where lifetime semantics of objects are not as obvious.
